internal/enricher/geolocation.go

```diff
@@ -0,0 +1,235 @@
+			"proc_geo_enricher_request",
+			stats.CountType,
+			stats.Tags{
+				"sourceID": source.ID,
+				"error":    errType,
+			}).Increment()
+	}()
+
+	if request.RequestIP == "" {
+		return nil
+	}
+
+	defer e.stats.NewTaggedStat(
+		"proc_geo_enricher_request_latency",
+		stats.TimerType,
+		stats.Tags{
+			"sourceID": source.ID,
+		},
+	).RecordDuration()()
+
+	rawGeo, err = e.fetcher.Locate(request.RequestIP)
+	if err != nil {
+		return fmt.Errorf("enriching request with geolocation: %w", err)
+	}
+
+	geoData := extractGeolocationData(rawGeo)
+
+	for _, event := range request.Batch {
+		// if the context section is missing on the event
+		// set it with default as map[string]interface{}
+		if _, ok := event["context"]; !ok {
+			event["context"] = map[string]interface{}{}
+		}
+
+		if context, ok := event["context"].(map[string]interface{}); ok {
+			// if the `geo` key already present fire a stat for observability
+			if _, ok := context["geo"]; ok {
+				continue
+			}
+			// update the geo section
+			context["geo"] = geoData
+		}
+	}
+
+	return nil
+}
+
+func (e *geoEnricher) Close() error {
+	if err := e.fetcher.Close(); err != nil {
+		return fmt.Errorf("closing the geo enricher: %w", err)
+	}
+	return nil
+}
+
+func downloadMaxmindDB(ctx context.Context, conf *config.Config, log logger.Logger) (string, error) {
+	var (
+		dbKey   = conf.GetString("Geolocation.db.key", "geolite2City.mmdb")
+		bucket  = conf.GetString("Geolocation.db.bucket", "rudderstack-geolocation")
+		region  = conf.GetString("Geolocation.db.bucket.region", "us-east-1")
+		timeout = conf.GetInt("Geolocation.db.download.timeout", 100)
+	)
+
+	log.Infof("downloading new geolocation db from key: %s", dbKey)
+	var (
+		tmpDirPath = strings.TrimSuffix(conf.GetString("RUDDER_TMPDIR", "."), "/")
+		baseDIR    = fmt.Sprintf("%s/geolocation", tmpDirPath)
+		fullpath   = fmt.Sprintf("%s/%s", baseDIR, dbKey)
+	)
+
+	// If the filepath exists return
+	_, err := os.Stat(fullpath)
+	if err == nil {
+		return fullpath, nil
+	}
+
+	err = os.MkdirAll(baseDIR, os.ModePerm)
+	if err != nil {
+		return "", fmt.Errorf("creating directory for storing db: %w", err)
+	}
+
+	f, err := os.CreateTemp(baseDIR, "geodb-*.mmdb")
+	if err != nil {
+		return "", fmt.Errorf("creating a temporary file: %w", err)
+	}
+
+	defer func() {
+		f.Close()
+		os.Remove(f.Name())
+	}()
+
+	manager, err := filemanager.NewS3Manager(map[string]interface{}{
+		"bucketName": bucket,
+		"region":     region,
+	}, log, func() time.Duration {
+		return time.Duration(timeout) * time.Second
+	})
+	if err != nil {
+		return "", fmt.Errorf("creating a new s3 manager client: %w", err)
+	}
+
+	err = manager.Download(ctx, f, dbKey)
+	if err != nil {
+		return "", fmt.Errorf("downloading file with key: %s from bucket: %s and region: %s, err: %w",
+			dbKey,
+			bucket,
+			region,
+			err)
+	}
+
+	// Finally move the downloaded file from previous temp location to new location
+	err = os.Rename(f.Name(), fullpath)
```

Comment:
Can we rename an open file without closing it? I believe it is about time to add some test coverage for this function. Starting a MINIO server and pushing a file to it is easy.

Reply:
I am confused a bit on this Aris, you mentioning to write tests for `downloadMaxmindDB` or for `os.Rename` separately ?

Reply:
The function